pkg/virt-controller/services/nodeselectorrenderer.go

```diff
@@ -0,0 +1,212 @@
+package services
+
+import (
+	"fmt"
+
+	v1 "kubevirt.io/api/core/v1"
+
+	"kubevirt.io/kubevirt/pkg/virt-controller/watch/topology"
+)
+
+type NodeSelectorRenderer struct {
+	cpuFeatureLabels []string
+	cpuModelLabel    string
+	hasDedicatedCPU  bool
+	hyperv           bool
+	podNodeSelectors map[string]string
+	tscFrequency     *int64
+	vmiFeatures      *v1.Features
+}
+
+type NodeSelectorRendererOption func(renderer *NodeSelectorRenderer)
+
+func NewNodeSelectorRenderer(
+	vmiNodeSelectors map[string]string,
+	clusterWideConfNodeSelectors map[string]string,
+	opts ...NodeSelectorRendererOption,
+) *NodeSelectorRenderer {
+	podNodeSelectors := map[string]string{v1.NodeSchedulable: "true"}
+	copySelectors(clusterWideConfNodeSelectors, podNodeSelectors)
+	copySelectors(vmiNodeSelectors, podNodeSelectors)
+
+	nodeSelectorRenderer := &NodeSelectorRenderer{podNodeSelectors: podNodeSelectors}
+	for _, opt := range opts {
+		opt(nodeSelectorRenderer)
+	}
+	return nodeSelectorRenderer
+}
+
+func (nsr *NodeSelectorRenderer) Render() map[string]string {
+	if nsr.hasDedicatedCPU {
+		nsr.enableSelectorLabel(v1.CPUManager)
+	}
+	if nsr.hyperv {
+		copySelectors(hypervNodeSelectors(nsr.vmiFeatures), nsr.podNodeSelectors)
+	}
+	if nsr.cpuModelLabel != "" && nsr.cpuModelLabel != cpuModelLabel(v1.CPUModeHostModel) && nsr.cpuModelLabel != cpuModelLabel(v1.CPUModeHostPassthrough) {
+		nsr.enableSelectorLabel(nsr.cpuModelLabel)
+	}
+	for _, cpuFeatureLabel := range nsr.cpuFeatureLabels {
+		nsr.enableSelectorLabel(cpuFeatureLabel)
+	}
+
+	if nsr.isTSCSchedulable() {
+		nsr.enableSelectorLabel(topology.ToTSCSchedulableLabel(*nsr.tscFrequency))
+	}
+
+	return nsr.podNodeSelectors
+}
+
+func (nsr *NodeSelectorRenderer) enableSelectorLabel(label string) {
+	nsr.podNodeSelectors[label] = "true"
+}
+
+func (nsr *NodeSelectorRenderer) isTSCSchedulable() bool {
+	return nsr.tscFrequency != nil
+}
+
+func WithDedicatedCPU() NodeSelectorRendererOption {
+	return func(renderer *NodeSelectorRenderer) {
+		renderer.hasDedicatedCPU = true
+	}
+}
+
+func copySelectors(src map[string]string, dst map[string]string) {
+	for k, v := range src {
+		dst[k] = v
+	}
+}
+
+func WithHyperv(features *v1.Features) NodeSelectorRendererOption {
+	return func(renderer *NodeSelectorRenderer) {
+		renderer.hyperv = true
+		renderer.vmiFeatures = features
+	}
+}
+
+func WithModelAndFeatureLabels(modelLabel string, cpuFeatureLabels ...string) NodeSelectorRendererOption {
+	return func(renderer *NodeSelectorRenderer) {
+		renderer.cpuFeatureLabels = cpuFeatureLabels
+		renderer.cpuModelLabel = modelLabel
+	}
+}
```

Comment:
That looks a bit strange\misleading because if the CPU-model of the vmi is defined and it's not `host-model`(default model) or `Passthrough` we don't  need selector  for the CPU-Features as it is a well defined model.